collect_app/src/main/java/org/odk/collect/android/formmanagement/formmap/FormMapViewModel.kt

```diff
@@ -55,29 +56,45 @@ class FormMapViewModel(
         return mappableItems
     }
 
+    override fun isLoading(): NonNullLiveData<Boolean> {
+        return isLoading
+    }
+
     fun load() {
-        val instances = instancesRepository.getAllByFormId(form.formId)
-        val items: MutableList<MappableSelectItem> = ArrayList()
-
-        for (instance in instances) {
-            if (instance.geometry != null && instance.geometryType == Instance.GEOMETRY_TYPE_POINT) {
-                try {
-                    val geometry = JSONObject(instance.geometry)
-                    val coordinates = geometry.getJSONArray("coordinates")
-
-                    // In GeoJSON, longitude comes before latitude.
-                    val lon = coordinates.getDouble(0)
-                    val lat = coordinates.getDouble(1)
-
-                    items.add(createItem(instance, lat, lon))
-                } catch (e: JSONException) {
-                    Timber.w("Invalid JSON in instances table: %s", instance.geometry)
+        isLoading.value = true
+
+        scheduler.immediate(
```

Comment:
Shouldn't this use `viewmodelscope`? As i understand scheduler.immediate which launches `CoroutineScope` won't be canceled when `viewmodel` is destroyed or maybe I'm missing something?

Reply:
We haven't actually used the new `viewModelScope` yet - our `Scheduler` interface doesn't support it. I'm not sure if we get a lot of benefit here, as all it would do is cancel the task if you leave the screen while the instances are loading. It's potentially wasting a bit of CPU time to keep going, but it shouldn't have any negative consequences otherwise, as the view code will be detached from the `LiveData` object that the results get put into anyway. 

Maybe we should create an issue to create support for `viewModelScope` in some way in `Scheduler` and then come back for this and other `Scheduler` uses in `ViewModel` classes. How does that sound?

Reply:
Ok we can have a separate issue. This case seems like a perfect example where `viewModelScope` should be used. Didn't it exist back then when you created that interface?

Reply:
Potentially! It was likely either undocumented or still alpha/beta. That, or I just missed it. It's important to remember that `viewModelScope` cancels coroutines on `ViewModel#onCleared` but only if the coroutine "cooperates" with being cancelled. By that, I mean we'd need to actually be checking `isActive` within the code to have cancellation avoid us doing extra work in the background. Without that, we're only really going to end up cancelling our `foreground` work, which will usually just amount to setting a value on `LiveData` (which should no longer be observed if the `ViewModel` has been destroyed). There's some good docs on this [here](https://kotlinlang.org/docs/cancellation-and-timeouts.html#cancellation-is-cooperative).

I guess in this specific case (and in anywhere the background task is "quick"), I'd argue that we wouldn't really gain much from cancellation. Does that make sense? Thinking about it some more. I'm feeling like we should only really be introducing specific scopes/lifecycle aware tasks for longer running jobs (form download would be a good example) and it might be that we want to move all that stuff to `WorkManager` anyway.